### PR TITLE
Refactor minmax API

### PR DESCRIPTION
**Pull request by steff456**
This PR is part of https://github.com/stdlib-js/todo/issues/1435

## Description

> What is the purpose of this pull request?

This pull request:

-   Refactors `minmax` package to support `assign` when the result needs to overwrite a given array

## Related Issues

> Does this pull request have any related issues?

This pull request is part of https://github.com/stdlib-js/todo/issues/1435

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

- The typescript linter doesn't like that the optional arguments are defined in the middle of the `assign` signature. 

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/math/base/special/minmax/benchmark/benchmark.js [hunk 1]
```diff
@@ -67,7 +67,7 @@ bench( pkg+'::memory_reuse', function benchmark( b ) {
 	for ( i = 0; i < b.iterations; i++ ) {
 		x = ( randu()*1000.0 ) - 500.0;
 		y = ( randu()*1000.0 ) - 500.0;
-		z = minmax( out, x, y );
+		z = minmax( x, y, out, 1, 0 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/benchmark/benchmark.js]: Need to update L57 from `::memory_reuse` to `pkg+':assign'`

File: lib/node_modules/@stdlib/math/base/special/minmax/docs/repl.txt [hunk 2]
```diff
@@ -37,11 +34,47 @@
...
+    Returns
+    -------
+    out: Array|TypedArray|Object
+        Minimum and maximum values.
+
+    Examples
+    --------
+    > var out = [ 0.0, 0.0 ];
+    > v = {{alias}}.assign( 3.14, out, 1, 0 )
     [ 3.14, 3.14 ]
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/docs/repl.txt]:
```suggestion
    > var v = {{alias}}.assign( 3.14, -1.5, out, 1, 0 )
    [ -1.5, 3.14 ]
```

Slightly more exciting example.

File: lib/node_modules/@stdlib/math/base/special/minmax/docs/types/index.d.ts [hunk 3]
```diff
@@ -23,23 +23,51 @@
...
+	*
+	* @example
+	* var out = [ 0.0, 0.0 ];
+	* var v = minmax( 5.9, 3.14, 4.2, out, 1, 0 );
+	* // returns [ 3.14, 5.9 ]
+	*
+	* var bool = ( v === out );
+	* // returns true
+	*/
+	assign( x: number, y?: number, ...args: Array<number>, out: ArrayLike<any>, stride: number, offset: number ): Collection; // tslint-disable-line max-line-length
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/docs/types/index.d.ts:69]: We need to use overloads here. I can do this on my end.

File: lib/node_modules/@stdlib/math/base/special/minmax/docs/types/test.ts [hunk 4]
```diff
@@ -53,9 +53,100 @@ import minmax = require( './index' );
...
-	minmax( false, 3 ); // $ExpectError
-	minmax( {}, 3 ); // $ExpectError
+	minmax(); // $ExpectError
+}
+
+// Attached to the main export is an `assign` method which returns an array-like object containing numbers...
+{
+	const out = [ 0.0, 0 ];
+
+	minmax.assign( 1.2, 2.2, [3, 4], out, 1, 0 ); // $ExpectType Collection
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/docs/types/test.ts]:
```suggestion
	minmax.assign( 1.2, 2.2, out, 1, 0 ); // $ExpectType Collection
```

File: lib/node_modules/@stdlib/math/base/special/minmax/docs/types/test.ts [hunk 5]
```diff
@@ -53,9 +53,100 @@ import minmax = require( './index' );
...
 {
-	minmax( true, 3 ); // $ExpectError
-	minmax( false, 3 ); // $ExpectError
-	minmax( {}, 3 ); // $ExpectError
+	minmax(); // $ExpectError
+}
+
+// Attached to the main export is an `assign` method which returns an array-like object containing numbers...
+{
+	const out = [ 0.0, 0 ];
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/docs/types/test.ts]:
```suggestion
	const out = [ 0.0, 0.0 ];
```

File: lib/node_modules/@stdlib/math/base/special/minmax/docs/types/test.ts [hunk 6]
```diff
@@ -53,9 +53,100 @@ import minmax = require( './index' );
...
+	minmax.assign( 1, 2, false, out, 1, 0 ); // $ExpectError
+	minmax.assign( 1, 2, true, out, 1, 0 ); // $ExpectError
+	minmax.assign( 1, 2, '2', out, 1, 0 ); // $ExpectError
+	minmax.assign( 1, 2, null, out, 1, 0 ); // $ExpectError
+	minmax.assign( 1, 2, ['3'], out, 1, 0 ); // $ExpectError
+	minmax.assign( 1, 2, {}, out, 1, 0 ); // $ExpectError
+	minmax.assign( 1, 2, ( x: number ): number => x, out, 1, 0 ); // $ExpectError
+}
+
+// The compiler throws an error if the `assign` method is provided a fourth argument which is not an array-like object of numbers...
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/docs/types/test.ts:107]: Once we add overloads, these tests will need to be updated.

File: lib/node_modules/@stdlib/math/base/special/minmax/lib/main.js [hunk 7]
```diff
@@ -20,75 +20,36 @@
...
-		i = 0;
-	} else {
-		args = [];
-		args.push( arguments[ 0 ] );
-		i = 1;
-	}
-	for ( ; i < len; i++ ) {
-		args.push( arguments[ i ] );
-	}
+	args = [].slice.call( arguments, 0 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/lib/main.js]: We need to explicitly copy arguments using a `for` loop. Passing `arguments` to a function can result in an automatic de-opt in older JS engines.

File: lib/node_modules/@stdlib/math/base/special/minmax/test/test.assign.js [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+	t.strictEqual( v[ 1 ], 4.2, 'returns max value' );
+
+	out = new Float64Array( 2 );
+	v = minmax( 4.2, 3.14, -1.0, -3.14, out, 1, 0 );
+	t.strictEqual( v, out, 'returns output array' );
+	t.strictEqual( v[ 0 ], -3.14, 'returns min value' );
+	t.strictEqual( v[ 1 ], 4.2, 'returns max value' );
+
+	t.end();
+});
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/test/test.assign.js:328]:
```suggestion
```

**Review by kgryte — COMMENTED** [review]
Overall, looks good! Just a few minor changes and I can update the TypeScript declarations on my end.

**Comment by steff456** [comment]
@kgryte the only thing that remains are the TypeScript declarations 😉

File: lib/node_modules/@stdlib/math/base/special/minmax/lib/main.js [hunk 9]
```diff
@@ -20,75 +20,40 @@
...
-		args.push( [ 0.0, 0.0 ] );
-		i = 0;
-	} else {
-		args = [];
-		args.push( arguments[ 0 ] );
-		i = 1;
-	}
-	for ( ; i < len; i++ ) {
+	args = [ x, y ];
+	for ( i=0; i < arguments.length; i++ ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/lib/main.js]: Should this start at `i = 2`?

File: lib/node_modules/@stdlib/math/base/special/minmax/benchmark/benchmark.js [hunk 10]
```diff
@@ -54,7 +54,7 @@ bench( pkg, function benchmark( b ) {
 	b.end();
 });
 
-bench( pkg+'::memory_reuse', function benchmark( b ) {
+bench( pkg+'::assign', function benchmark( b ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/benchmark/benchmark.js]:
```suggestion
bench( pkg+':assign', function benchmark( b ) {
```

File: lib/node_modules/@stdlib/math/base/special/minmax/docs/repl.txt [hunk 11]
```diff
@@ -36,12 +33,47 @@
     [ -0.0, +0.0 ]
     > v = {{alias}}( 3.14 )
     [ 3.14, 3.14 ]
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/docs/repl.txt:36]: L21 should be:

```
out: Array<number>
```

File: lib/node_modules/@stdlib/math/base/special/minmax/docs/repl.txt [hunk 12]
```diff
@@ -36,12 +33,47 @@
...
+    -------
+    out: Array|TypedArray|Object
+        Minimum and maximum values.
+
+    Examples
+    --------
     > var out = [ 0.0, 0.0 ];
-    > v = {{alias}}( out, 3.14 )
-    [ 3.14, 3.14 ]
+    > v = {{alias}}.assign( 3.14, -1.5, out, 1, 0 )
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/docs/repl.txt]:
```suggestion
    > var v = {{alias}}.assign( 3.14, -1.5, out, 1, 0 )
```

File: lib/node_modules/@stdlib/math/base/special/minmax/docs/types/index.d.ts [hunk 13]
```diff
@@ -23,23 +23,51 @@
...
+	* // returns [ 3.14, 4.2 ]
+	*
+	* var v = minmax( 3.14, NaN );
+	* // returns [ NaN, NaN ]
+	*
+	* @example
+	* var v = minmax( +0.0, -0.0 );
+	* // returns [ -0.0, 0.0 ]
+	*/
+	( x: number, y?: number, ...args: Array<number> ): Collection;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/docs/types/index.d.ts]:
```suggestion
	( x: number, y?: number, ...args: Array<number> ): Array<number>;
```

File: lib/node_modules/@stdlib/math/base/special/minmax/docs/types/test.ts [hunk 14]
```diff
@@ -26,10 +26,10 @@ import minmax = require( './index' );
 	minmax( -0.2 ); // $ExpectType Collection
 	minmax( 3, -0.2 ); // $ExpectType Collection
 	minmax( 3, -0.2, -1.2, -4.0 ); // $ExpectType Collection
-	minmax( [], -4.0 ); // $ExpectType Collection
+	minmax( -4.0 ); // $ExpectType Collection
 }
 
-// The compiler throws an error if the function is provided non-numbers for all but the first argument...
+// The compiler throws an error if the function is provided non-numbers for all arguments...
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/docs/types/test.ts]:
```suggestion
// The compiler throws an error if the function is provided non-numbers for any argument...
```

File: lib/node_modules/@stdlib/math/base/special/minmax/docs/types/test.ts [hunk 15]
```diff
@@ -26,10 +26,10 @@ import minmax = require( './index' );
 	minmax( -0.2 ); // $ExpectType Collection
 	minmax( 3, -0.2 ); // $ExpectType Collection
 	minmax( 3, -0.2, -1.2, -4.0 ); // $ExpectType Collection
-	minmax( [], -4.0 ); // $ExpectType Collection
+	minmax( -4.0 ); // $ExpectType Collection
 }
 
-// The compiler throws an error if the function is provided non-numbers for all but the first argument...
+// The compiler throws an error if the function is provided non-numbers for any argument...
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/docs/types/test.ts]:
```suggestion
// The compiler throws an error if the function is provided value which is not a number for any argument...
```

File: lib/node_modules/@stdlib/math/base/special/minmax/docs/types/test.ts [hunk 16]
```diff
@@ -26,10 +26,10 @@ import minmax = require( './index' );
 	minmax( -0.2 ); // $ExpectType Collection
 	minmax( 3, -0.2 ); // $ExpectType Collection
 	minmax( 3, -0.2, -1.2, -4.0 ); // $ExpectType Collection
-	minmax( [], -4.0 ); // $ExpectType Collection
+	minmax( -4.0 ); // $ExpectType Collection
 }
 
-// The compiler throws an error if the function is provided non-numbers for all but the first argument...
+// The compiler throws an error if the function is provided value which is not a number for any argument...
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/docs/types/test.ts]:
```suggestion
// The compiler throws an error if the function is provided an argument which is not a number...
```

File: lib/node_modules/@stdlib/math/base/special/minmax/lib/main.js [hunk 17]
```diff
@@ -20,75 +20,40 @@
...
 
 /**
 * Returns the minimum and maximum values.
 *
-* @param {(Array|TypedArray|Object)} [out] - output object
 * @param {number} x - first number
 * @param {number} [y] - second number
 * @param {...number} [args] - numbers
-* @returns {(Array|TypedArray|Object)} minimum and maximum values
+* @returns {Collection} minimum and maximum values
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/lib/main.js]:
```suggestion
* @returns {Array<number>} minimum and maximum values
```

File: lib/node_modules/@stdlib/math/base/special/minmax/docs/repl.txt (hunk 11 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/minmax/docs/repl.txt:36]: @steff456 I think this is the main thing remaining to be done.

File: lib/node_modules/@stdlib/math/base/special/minmax/benchmark/benchmark.js (hunk 1 above)
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/benchmark/benchmark.js]:
```suggestion
		z = minmax.assign( x, y, out, 1, 0 );
```

File: lib/node_modules/@stdlib/math/base/special/minmax/docs/repl.txt [hunk 18]
```diff
@@ -36,12 +33,47 @@
...
+    x: number
+        First number.
+
+    y: number (optional)
+        Second number.
+
+    args: ...number (optional)
+        Numbers.
+
+    out: Array<number>
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/docs/repl.txt]:
```suggestion
    out: Array|TypedArray|Object
```

File: lib/node_modules/@stdlib/math/base/special/minmax/docs/repl.txt [hunk 19]
```diff
@@ -36,12 +33,47 @@
...
+
+    stride: integer
+        Output array stride.
+
+    offset: integer
+        Output array index offset.
+
+    Returns
+    -------
+    out: Array<number>
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/docs/repl.txt]:
```suggestion
    out: Array|TypedArray|Object
```

File: lib/node_modules/@stdlib/math/base/special/minmax/lib/main.js [hunk 20]
```diff
@@ -20,75 +20,40 @@
...
-		args.push( [ 0.0, 0.0 ] );
-		i = 0;
-	} else {
-		args = [];
-		args.push( arguments[ 0 ] );
-		i = 1;
-	}
-	for ( ; i < len; i++ ) {
+	args = [ x, y ];
+	for ( i=0; i < arguments.length; i++ ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/minmax/lib/main.js]:
```suggestion
	for ( i = 0; i < arguments.length; i++ ) {
```

**Review by kgryte — APPROVED** [review]
Will go ahead and merge. We'll work on the TypeScript declarations in a follow-up commit.

**Pull request merged**